personhood/user/builder.go

```diff
@@ -0,0 +1,284 @@
+package user
+
+import (
+	"go.dedis.ch/cothority/v3/byzcoin"
+	contracts2 "go.dedis.ch/cothority/v3/byzcoin/contracts"
+	"go.dedis.ch/cothority/v3/darc"
+	"go.dedis.ch/cothority/v3/darc/expression"
+	"go.dedis.ch/cothority/v3/personhood/contracts"
+	"go.dedis.ch/kyber/v3/util/random"
+	"go.dedis.ch/protobuf"
+	"golang.org/x/xerrors"
+	"time"
+)
+
+// Builder allows to create a new user either directly from a DARC with
+// the appropriate permissions, or from a Spawner contract.
+type Builder struct {
+	credentialStruct contracts.CredentialStruct
+	darcDevice       *darc.Darc
+	darcSigner       *darc.Darc
+	darcCoin         *darc.Darc
+	darcCred         *darc.Darc
+	alias            string
+	coinID           byzcoin.InstanceID
+	spawnerID        byzcoin.InstanceID
+	signer           darc.Signer
+	UserID           byzcoin.InstanceID
+}
+
+// NewUserBuilder creates a new Builder.
+func NewUserBuilder(alias string) (*Builder, error) {
+	ub := &Builder{
+		alias:            alias,
+		signer:           darc.NewSignerEd25519(nil, nil),
+		credentialStruct: contracts.CredentialStruct{},
+	}
+	ub.SetAlias(alias)
+	ub.credentialStruct.SetConfig(contracts.ACView, []byte("c4dt_user"))
+	return ub, nil
+}
+
+// CreateFromDarc creates a new user from a darc that has the necessary
+// rules to create all elements.
+func (ub *Builder) CreateFromDarc(cl *byzcoin.Client, spawnerDarcID darc.ID,
+	spawnerSigner darc.Signer) (*User, error) {
+	if err := ub.createDarcs(); err != nil {
+		return nil, xerrors.Errorf("creating darcs: %v", err)
+	}
+	instrs, err := ub.instrsFromDarc(spawnerDarcID)
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't create instructions: %v", err)
+	}
+	instrs = append(instrs, ub.createCoin(spawnerDarcID)...)
+
+	return ub.createUser(cl, instrs, spawnerSigner)
+}
+
+// CreateFromSpawner creates a new user from a spawner instance.
+func (ub *Builder) CreateFromSpawner(spawner ActiveSpawner) (*User, error) {
+	ub.SetSpawnerID(spawner.iid)
+	if err := ub.createDarcs(); err != nil {
+		return nil, xerrors.Errorf("creating darcs: %v", err)
+	}
+	if err := spawner.SpawnDarcs(*ub.darcCoin, *ub.darcCred, *ub.darcSigner,
+		*ub.darcDevice); err != nil {
+		return nil, xerrors.Errorf("couldn't spawn darcs: %v", err)
+	}
+
+	coinIID, err := spawner.SpawnCoin(contracts2.CoinName,
+		ub.darcCoin.GetBaseID(), 10000)
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't spawn coin: %v", err)
+	}
+	ub.SetCoinID(coinIID)
+
+	credIID, err := spawner.SpawnCredential(ub.credentialStruct,
+		ub.darcCred.GetBaseID())
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't spawn credential: %v", err)
+	}
+	ub.UserID = credIID
+
+	if err := spawner.SendTransaction(); err != nil {
+		return nil, xerrors.Errorf("couldn't send transaction: %v", err)
+	}
+
+	time.Sleep(time.Second)
```

Comment:
better to avoid sleeping, or at least use the block time.

Reply:
Should be a `WaitForPropagation`